packages/require/lib/index.ts

```diff
@@ -0,0 +1,130 @@
+import path from "path";
+import TruffleConfig from "@truffle/config";
+import { options as expectOptions } from "@truffle/expect";
+import { Web3Shim, createInterfaceAdapter } from "@truffle/interface-adapter";
+
+import vm from "vm";
+import Module from "module";
+
+import originalRequire from "original-require";
+
+import { ExecOptions, RequireOptions } from "./types";
+import { compile, registerTsNode } from "./compiler";
+
+export function file(options: RequireOptions) {
+  const scriptPath = path.resolve(options.file);
+
+  expectOptions(options, ["file"]);
+
+  const conf = TruffleConfig.default().with(options);
+
+  const scriptModule = new Module(scriptPath);
+
+  // Provide all the globals listed here: https://nodejs.org/api/globals.html
+  const sandbox: { [index: string]: any } = {
+    __filename: path.basename(scriptPath),
+    __dirname: path.dirname(scriptPath),
+    Buffer,
+    clearImmediate,
+    clearInterval,
+    clearTimeout,
+    console,
+    exports: scriptModule.exports,
+    global,
+    process,
+    setImmediate,
+    setInterval,
+    setTimeout,
+    module: scriptModule,
+    config: conf,
+    artifacts: conf.resolver,
+    require: (id: string) => {
+      // Ugh. Simulate a full require function for the file.
+      id = id.trim();
+
+      // If absolute, just require.
+      if (path.isAbsolute(id)) return originalRequire(id);
+
+      // If relative, it's relative to the file.
+      if (id[0] === ".") {
+        return originalRequire(path.join(path.dirname(scriptPath), id));
+      } else {
+        // Not absolute, not relative, must be a globally or locally installed module.
+        // Try local first.
+        // Here we have to require from the node_modules directory directly.
+
+        var moduleDir = conf.working_directory;
+        while (true) {
+          try {
+            return originalRequire(path.join(moduleDir, "node_modules", id));
+          } catch (e) {}
+          var oldModuleDir = moduleDir;
+          moduleDir = path.dirname(moduleDir);
+          if (moduleDir === oldModuleDir) break;
+        }
+
+        // Try global, and let the error throw.
+        return originalRequire(id);
+      }
+    }
+  };
+
+  // Now add contract names.
+  Object.keys(conf.context || {}).forEach(key => {
+    sandbox[key] = conf.context[key];
+  });
+
+  const context = vm.createContext(sandbox);
+
+  const old_cwd = process.cwd();
+  const cwd = path.dirname(scriptPath);
+  process.chdir(cwd);
+
+  registerTsNode(conf, scriptPath, context);
```

Comment:
I think this chunk of code is probably the lowest-hanging fruit as far as performance improvement goes, but I left it alone for now to preserve behavior for JS code.

That is, to make this faster we could cache the script global context and only run `registerTsNode` once for the lifetime of that context.

Reply:
I wonder if you could do something quick and easy like `if (!context.process[Symbol.for("ts-node.register.instance")]) { registerTsNode(conf, scriptPath, context); }`?

Reply:
I think that once contextified the `context` needs to be treated like an opaque type. But that said, `context.process` should be the same as `process`, which makes me wonder if all of these hoops that I'm jumping through to only register it in the child context are for nought.

Reply:
I changed up the flow of this and now we only `registerTsNode` one time.